### PR TITLE
fixing autodiff to support Optional[Tensor] on inputs

### PR DESCRIPTION
**Pull request by jjsjann123**
This PR fixes two local issue for me:

1. Assert failure when passing `None` to `Optional[Tensor]` input that requires gradient in autodiff
2. Wrong vjp mapping on inputs when `requires_grad` flag changes on inputs stack.

This PR is to support autodiff on layer_norm.

### CODE REVIEW

**Comment by t-vi** [comment]
@jjsjann123 Do you have an immediate example that we could use for a test? If not I'd try to come up with one.

**Comment by jjsjann123** [comment]
I can stack another PR (enable layer_norm for autodiff) with this one for testing. Right now we don't have `Optional[Tensor]` in autodiff that takes gradient, so I don't know how to get a reasonable example :(

**Comment by t-vi** [comment]
Nono. That's cool.

File: torch/csrc/jit/runtime/graph_executor.cpp [hunk 1]
```diff
@@ -299,7 +300,12 @@ struct DifferentiableGraphBackward : public autograd::Node {
         addOutputForTensor(tensor);
       }
     } else {
-      addOutputForTensor(value.toTensor());
+      if (value.isTensor()) {
```
Comment (eellison) [inline torch/csrc/jit/runtime/graph_executor.cpp:303]: nit: `else ... if () ->  elif`

File: torch/csrc/jit/runtime/graph_executor.cpp [hunk 2]
```diff
@@ -299,7 +300,12 @@ struct DifferentiableGraphBackward : public autograd::Node {
...
+        addOutputForTensor(value.toTensor());
+      } else {
+        // We could have None passed here via `Optional[Tensor]`
+        add_next_edge(autograd::Edge{});
```
Comment (eellison) [inline torch/csrc/jit/runtime/graph_executor.cpp:307]: this lgtm,  but ping @albanD just to be safe

**Review by eellison — COMMENTED** [review]
Nice,  LGTM! Mind adding a test? @albanD mind taking a look a the autograd::Edge line ?

File: torch/csrc/jit/runtime/graph_executor.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/jit/runtime/graph_executor.cpp:307]: That looks good. Especially because it matches what is done in `addOutputForTensor` just above.

**Comment by jjsjann123** [comment]
> Mind adding a test?

Can you point me to an example for a functional cpp test on this class? Doesn't look obvious to me how to do an isolated test on this.
I did verify the fix in a separate python test. But that was with a stacked PR with layer_norm enabled in autodiff (We'll issue that one when this is merged). Right now we don't have autodiff supporting gradient on optional tensor so I can't mimic that test here in this PR.

**Comment by eellison** [comment]
Yea that’s fine, we can test with layer norm PR. I didn’t see the earlier convo abt this.

**Review by eellison — APPROVED** [review]
